lib/semantics/canonicalize-do.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "canonicalize-do.h"
+#include "../parser/parse-tree-visitor.h"
+
+namespace Fortran::semantics {
+
+class CanonicalizationOfDoLoops {
+public:
+  CanonicalizationOfDoLoops() = default;
+
+  template<typename T> bool Pre(T &) { return true; }
+  template<typename T> void Post(T &) {}
+  bool Pre(parser::ExecutionPart &executionPart) {
+    const auto &endIter{executionPart.v.end()};
+    currentList_ = &executionPart.v;
+    for (auto iter{executionPart.v.begin()}; iter != endIter; ++iter) {
+      iter = CheckStatement(iter);
```

Comment:
Please be sure that you want to both update `iter` with `CheckStatement` *and* increment it in the `for` statement's increment expression.  If `CheckStatement` really just checks statements, I have no idea what its returned iterator might be pointing at.

Reply:
yes, it is necessary to keep the iterator pointing at a valid list member.